ssm/message_directory.py

```diff
@@ -0,0 +1,127 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the MessageDirectory class."""
+
+import logging
+import os
+import uuid
+
+# logging configuration
+log = logging.getLogger(__name__)
+
+
+class MessageDirectory(object):
+    """A structure for holding Accounting messages in a directory."""
+
+    def __init__(self, path):
+        """Create a new directory structure for holding Accounting messages."""
+        self.directory_path = path
+
+    def add(self, data):
+        """Add the passed data to a new file and return it's name."""
+        # Create a unique file name so APEL admins can pair sent and recieved
+        # messages easily (as the file name appears in the sender and receiver
+        # logs as the message ID).
+        name = uuid.uuid4()
+
+        # Open the file and write the provided data into the file.
+        with open("%s/%s" % (self.directory_path, name), 'w') as message:
+            message.write(data)
+
+        # Return the name of the created file as a string,
+        # to keep the dirq like interface.
+        return "%s" % name
+
+    def count(self):
+        """
+        Return the number of elements in the queue.
+
+        Regardless of their state.
+        """
+        return len(self._get_messages())
+
+    def get(self, name):
+        """Return the content of the named message."""
+        with open("%s/%s" % (self.directory_path, name)) as message:
+            content = message.read()
+        return content
+
+    def lock(self, _name):
+        """Return True to simulate a successful lock. Does nothing else."""
+        return True
+
+    def purge(self):
+        """
+        Do nothing, as there are no old/intermediate directories to purge.
+
+        Only included to preserve dirq interface.
+        """
+        log.debug("purge called, but purge does nothing for non-dirq sending.")
+
+    def remove(self, name):
+        """Remove the named message."""
+        os.unlink("%s/%s" % (self.directory_path, name))
+
+    def _get_messages(self, sort_by_mtime=False):
+        """
+        Get the messages stored in this MessageDirectory.
+
+        if sort_by_mtime is set to True, the returned list is guaranteed to be
+        in increasing order of modification time.
+
+        mtime is used because (apparently) there is not way to find the
+        original date of file creation due to a limitation
+        of the underlying filesystem
```

Comment:
It's a POSIX limitation (see https://lwn.net/Articles/397442/), Windows does provide a creation time thouh, but then allows it to be modified, which probably makes it useless.

If you create the messages though, why not encode the creation time in their filename?

Reply:
We don't want any particular format of name to be a requirement as the point of this change is to allow the option of arbitrary files to be sent by SSM without using the `dirq` module and it's finicky naming scheme. We thought that sorting files by age before sending would at least give it some rough FIFO-ness.